dm/pkg/checker/privilege.go

```diff
@@ -122,72 +135,97 @@ func (pc *SourceReplicatePrivilegeChecker) Name() string {
 		}
 		grantStmt, ok := node.(*ast.GrantStmt)
 		if !ok {
 			switch node.(type) {
 			case *ast.GrantProxyStmt, *ast.GrantRoleStmt:
 				continue
 			default:
-				result.Errors = append(result.Errors, NewError("%s is not grant statement", grants[i]))
-				return
+				return NewError("%s is not grant statement", grants[i])
 			}
 		}
 
 		if len(grantStmt.Users) == 0 {
-			result.Errors = append(result.Errors, NewError("grant has no user %s", grantStmt.Text()))
-			return
-		} else if user == "" {
-			// show grants will only output grants for requested user
-			user = grantStmt.Users[0].User.Username
+			return NewError("grant has no user %s", grantStmt.Text())
 		}
 
-		for _, privElem := range grantStmt.Privs {
-			if privElem.Priv == mysql.AllPriv {
-				if grantStmt.Level.Level == ast.GrantLevelGlobal {
-					result.State = StateSuccess
-					return
-				}
-				// REPLICATION CLIENT, REPLICATION SLAVE, RELOAD should be global privileges,
-				// thus a non-global GRANT ALL is not enough
-				for expectedGrant := range lackGrants {
-					if _, ok := privNeedGlobal[expectedGrant]; !ok {
-						delete(lackGrants, expectedGrant)
+		dbName := grantStmt.Level.DBName
+		tableName := grantStmt.Level.TableName
+		switch grantStmt.Level.Level {
+		case ast.GrantLevelGlobal:
+			for _, privElem := range grantStmt.Privs {
+				switch privElem.Priv {
+				case mysql.AllPriv:
+					return nil
+				default:
+					// mysql> show master status;
+					// ERROR 1227 (42000): Access denied; you need (at least one of) the SUPER, REPLICATION CLIENT privilege(s) for this operation
+					if privElem.Priv == mysql.SuperPriv {
+						delete(lackGrants, mysql.ReplicationClientPriv)
 					}
+					delete(lackGrants, privElem.Priv)
 				}
-			} else if _, ok := lackGrants[privElem.Priv]; ok {
-				// check every privilege and remove it from expectedGrants
-				if _, ok := privNeedGlobal[privElem.Priv]; ok {
-					if grantStmt.Level.Level == ast.GrantLevelGlobal {
+			}
+		case ast.GrantLevelDB:
+			for _, privElem := range grantStmt.Privs {
+				switch privElem.Priv {
+				case mysql.SelectPriv, mysql.LockTablesPriv:
+					if _, ok := lackGrants[privElem.Priv]; !ok {
+						continue
+					}
+					if _, ok := lackGrants[privElem.Priv][dbName]; !ok {
+						continue
+					}
+					// currently, only SELECT privilege goes here. we didn't require SELECT to be granted globally,
+					// dumpling could report error if an allow-list table is lack of privilege.
+					// we only check that SELECT is granted on all columns, otherwise we can't SHOW CREATE TABLE
+					if len(privElem.Cols) != 0 {
+						continue
+					}
+					delete(lackGrants[privElem.Priv], dbName)
+					if len(lackGrants[privElem.Priv]) == 0 {
 						delete(lackGrants, privElem.Priv)
 					}
-				} else {
+				}
+			}
+		case ast.GrantLevelTable:
+			for _, privElem := range grantStmt.Privs {
+				switch privElem.Priv {
+				case mysql.SelectPriv, mysql.LockTablesPriv:
```

Comment:
in fact LockTablesPriv is not GrantLevelTable

https://dev.mysql.com/doc/refman/8.0/en/grant.html Table 13.11